Wikipedia/assets/codemirror/codemirror-index.html

```diff
@@ -1073,7 +1258,62 @@
           isAtLineEnd: editor.getLine(cursorPosition.line).length === cursorPosition.ch
         }
       },
-      newlineAndIndent: () => editor.execCommand('newlineAndIndent')
+      newlineAndIndent: () => editor.execCommand('newlineAndIndent'),
+      getLink: () => stateStore.link,
+      insertOrEditLink: (page, label) => {
+        const link = stateStore.link
+        if (!link) {
+          console.warn("Attempting to insert or edit link with no link in stateStore")
+          return
+        }
+        const position = link.pagePosition
+        
+        if (link.hasMarkup) { // existing link
+          if (link.label) {
+            if (label) { // modify an exsiting label
+              const pageWithLabel = `${page}|${label}`
+              editor.replaceRange(pageWithLabel, link.pagePosition.from, link.pagePosition.to) 
+            } else { // modify an existing link with a label
+              const pageWithExistingLabel = `${page}|${link.label}`
+              editor.replaceRange(pageWithExistingLabel, link.pagePosition.from, link.pagePosition.to)
+            }
+          } else {
+            if (label) { // add a label to an existing link with no label
+              const pageWithLabel = `${page}|${label}`
+              editor.replaceRange(pageWithLabel, link.pagePosition.from, link.pagePosition.to)
+            } else { // modify existing link
+              editor.replaceRange(page, link.pagePosition.from, link.pagePosition.to)
+            }
+          }
+        } else { // new link
+          if (label) {
+            const newPageWithLabel = `[[${page}|${label}]]`
+            editor.replaceRange(newPageWithLabel, link.pagePosition.from, link.pagePosition.to)
+          } else {
+            const newPage = `[[${page}]]`
+            editor.replaceRange(newPage, link.pagePosition.from, link.pagePosition.to)
+          }
```

Comment:
This new link block doesn't seem to be working. `link.pagePosition` is undefined. Should these be `replaceSelection` instead of `replaceRange`?

Reply:
Repro steps are select unlinked text (like "home"), tap the link button, tap an article

Reply:
`replaceSelection` won't work for 0 length selection